source/lac/psblas_vector.cc

```diff
@@ -0,0 +1,681 @@
+      }
+
+    // Create a new PSBLAS vector and allocate mem space for vector
+    psblas_vector = psb_c_new_dvector();
+
+    err = psb_c_dgeall_remote(psblas_vector, psblas_descriptor.get());
+    Assert(err == 0, internal::ExcInitializePSBLASVector(err));
+
+    if (omit_zeroing_entries == false)
+      {
+        err = psb_c_dvect_set_scal(psblas_vector, 0.0);
+        Assert(err == 0,
+               internal::ExcCallingPSBLASFunction(err, "psb_c_dvect_set_scal"));
+      }
+    state       = internal::State::Build;
+    last_action = VectorOperation::unknown;
+  }
+
+
+
+  void
+  Vector::reinit(const IndexSet &local_partitioning,
+                 const IndexSet &ghosts,
+                 const MPI_Comm  comm)
+  {
+    Assert(comm != MPI_COMM_NULL,
+           ExcMessage("MPI_COMM_NULL passed to Vector::reinit()."));
+    communicator   = comm;
+    ghosted        = true;
+    owned_elements = local_partitioning;
+
+    ghost_indices = ghosts;
+    ghost_indices.subtract_set(local_partitioning);
+
+    int err;
+    if (psblas_descriptor.get() == nullptr)
+      {
+        psblas_descriptor = std::shared_ptr<psb_c_descriptor>(
+          psb_c_new_descriptor(), internal::PSBLASDescriptorDeleter{});
+
+        // Use get_index_vector() from IndexSet to get the indexes
+        const std::vector<types::global_dof_index> &indexes =
+          local_partitioning.get_index_vector();
+
+        psb_i_t number_of_local_indexes =
+          indexes.size(); // Number of local indexes
+        // Copy the indexes into a psb_l_t array called vl
+        psb_l_t *vl =
+          (psb_l_t *)malloc(number_of_local_indexes * sizeof(psb_l_t));
+        psb_i_t *lidx =
+          (psb_i_t *)malloc(number_of_local_indexes * sizeof(psb_i_t));
+
+
+        for (psb_i_t i = 0; i < number_of_local_indexes; ++i)
+          {
+            vl[i]   = static_cast<psb_l_t>(indexes[i]);
+            lidx[i] = i;
+          }
+
+        // Ghost case. From the manual:
+        // 1) psb_cdall() with vl (global) and lidx (local)
+        // 2) psb_cdins(nz,ja,desc,info,lidx=lidx), where:
+        // - ja contains halo indices
+        // - lidx: corresponding local indices
+
+        // Insert the indexes into the descriptor
+        psblas_context = InitFinalize::get_psblas_context();
+        err            = psb_c_cdall_vl_lidx(number_of_local_indexes,
+                                  vl,
+                                  lidx,
+                                  *psblas_context,
+                                  psblas_descriptor.get());
+
+        // Free vl array
+        free(vl);
+        free(lidx);
+        Assert(err == 0, internal::ExcInitializePSBLASDescriptor(err));
+
+        // ... insert the ghost indices ...
+        const std::vector<types::global_dof_index> &ghost_indexes =
+          ghost_indices.get_index_vector();
+        const int number_of_ghost_indices = ghost_indexes.size();
```

Comment:
```suggestion
        const auto number_of_ghost_indices = ghost_indexes.size();
```
here and elsewhere we shouldn't convert to narrower integer types (I think the CI will also catch this)